src/main/java/gregtech/common/metatileentities/multi/multiblockpart/appeng/MetaTileEntityMEInputHatch.java

```diff
@@ -0,0 +1,374 @@
+import net.minecraft.client.resources.I18n;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.FluidTankInfo;
+import net.minecraftforge.fluids.IFluidTank;
+import net.minecraftforge.fluids.capability.FluidTankProperties;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * @Author GlodBlock
+ * @Description The Input Hatch that can auto fetch fluid ME storage network.
+ * @Date 2023/4/20-21:21
+ */
+public class MetaTileEntityMEInputHatch extends MetaTileEntityAEHostablePart implements IMultiblockAbilityPart<IFluidTank> {
+
+    public final static String FLUID_BUFFER_TAG = "FluidTanks";
+    public final static String WORKING_TAG = "WorkingEnabled";
+    private final static int CONFIG_SIZE = 16;
+    private boolean workingEnabled;
+    private ExportOnlyAEFluid[] aeFluidTanks;
+
+    public MetaTileEntityMEInputHatch(ResourceLocation metaTileEntityId) {
+        super(metaTileEntityId, GTValues.UHV, false);
+        this.workingEnabled = true;
+    }
+
+    @Override
+    protected void initializeInventory() {
+        this.aeFluidTanks = new ExportOnlyAEFluid[CONFIG_SIZE];
+        for (int i = 0; i < CONFIG_SIZE; i ++) {
+            this.aeFluidTanks[i] = new ExportOnlyAEFluid(null, null, this.getController());
+        }
+        super.initializeInventory();
+    }
+
+    @Override
+    protected FluidTankList createImportFluidHandler() {
+        return new FluidTankList(false, this.aeFluidTanks);
+    }
+
+    @Override
+    public void update() {
+        super.update();
+        if (!getWorld().isRemote && this.workingEnabled && this.shouldSyncME()) {
+            if (this.updateMEStatus()) {
+                try {
+                    IMEMonitor<IAEFluidStack> aeNetwork = this.getProxy().getStorage().getInventory(FLUID_NET);
+                    for (ExportOnlyAEFluid aeTank : this.aeFluidTanks) {
+                        // Try to clear the wrong fluid
+                        IAEFluidStack exceedFluid = aeTank.exceedStack();
+                        if (exceedFluid != null) {
+                            long total = exceedFluid.getStackSize();
+                            IAEFluidStack notInserted = aeNetwork.injectItems(exceedFluid, Actionable.MODULATE, this.getActionSource());
```

Comment:
no simulation before real insertion ?

Reply:
i don't think it needs simulation because it tells how much item/fluid isn't inserted, so i can change the uninserted item/fluid according to it. simulation doesn't make sense here